core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -330,6 +344,25 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
       completed
     }
 
+    // traverse the list and try to complete some watched elements
+    def tryCompleteFirstWatched(): Boolean = {
```

Comment:
I think the difficulty with this might be ensuring that all the watchers that were awaiting a particular key _eventually_ get notified. I think I see the problem you are trying to address. You don't want to have races between two watchers for the same key, but as long as we are protecting state transitions with a lock, I wonder how serious the concern is. Perhaps I'm missing something?